src/rust/aggregator/src/pipelines/order_history.rs

```diff
@@ -0,0 +1,324 @@
+        struct Fill {
+            market_id: BigDecimal,
+            maker_order_id: BigDecimal,
+            taker_order_id: BigDecimal,
+            txn_version: BigDecimal,
+            event_idx: BigDecimal,
+            size: BigDecimal,
+        }
+        let fills = sqlx::query_file_as!(
+            Fill,
+            "sqlx_queries/order_history/get_fill_events_between_txn_versions.sql",
+            txn_version_of_state,
+            txn_version
+        )
+        .fetch_all(transaction as &mut PgConnection)
+        .await
+        .map_err(to_pipeline_error)?;
+
+        // Get all change events that happened between the last transaction included in the previous
+        // state and the last transaction that happened before the given timestamp.
+        struct Change {
+            market_id: BigDecimal,
+            order_id: BigDecimal,
+            txn_version: BigDecimal,
+            event_idx: BigDecimal,
+            new_size: BigDecimal,
+        }
+        let changes = sqlx::query_file_as!(
+            Change,
+            "sqlx_queries/order_history/get_change_order_size_events_between_txn_versions.sql",
+            txn_version_of_state,
+            txn_version
+        )
+        .fetch_all(transaction as &mut PgConnection)
+        .await
+        .map_err(to_pipeline_error)?;
+
+        // Get all cancel events that happened between the last transaction included in the previous
+        // state and the last transaction that happened before the given timestamp.
+        let cancels = sqlx::query_file!(
+            "sqlx_queries/order_history/get_cancel_order_events_between_txn_versions.sql",
+            txn_version_of_state,
+            txn_version
+        )
+        .fetch_all(transaction as &mut PgConnection)
+        .await
+        .map_err(to_pipeline_error)?;
+
+        // Insert all places as orders into the state.
+        for place in places {
+            orders.insert((place.market_id.clone(), place.order_id.clone()), place);
+        }
+
+        // Remove all orders that have been cancelled.
+        for cancel in cancels {
+            orders.remove(&(cancel.market_id, cancel.order_id));
+        }
+
+        // Handle fills and changes chronologically.
+        let mut fills_index = 0;
+        let mut changes_index = 0;
+        let handle_fill =
+            |fill: &Fill, orders: &mut HashMap<OrderKey, Order>, index: &mut usize| {
+                *index += 1;
+                let order_key_maker = (fill.market_id.clone(), fill.maker_order_id.clone());
+                if let Some(order) = orders.get_mut(&order_key_maker) {
+                    order.size -= fill.size.clone();
+                    if order.size.is_zero() {
+                        orders.remove(&order_key_maker);
+                    }
+                }
+                let order_key_taker = (fill.market_id.clone(), fill.taker_order_id.clone());
+                if let Some(order) = orders.get_mut(&order_key_taker) {
```

Comment:
This only gets called in the case of a limit order that takes, correct?

Reply:
This closure gets called for every deduped fill. The two `if`s are here to ensure we did not already cancel the order in the `for` above.